src/ase/cw/model/Server.java

```diff
@@ -81,12 +91,27 @@ public void startOrderProcess() {
         } else {
             System.out.println(getName() + "already started");
         }
+
+    }
+
+    @Override
+    public void pauseOrderProcess() {
+        // this does not pause the status immediately but waits until current order is processed
+        serverStatus = status.PAUSED;
+    }
+
+    @Override
+    public void restartOrderProcess() {
+        synchronized (serverThread) {
+            serverThread.notify();
+        }
+        setStatus(status.FREE);
```

Comment:
You call setStatus in multiple threads. This can lead to raceConditions.
Please only update the ServerStatus inside the run() method.
You can use a internal helper variable to detect then the thread should pause. So only one thread change a variable value(Otherwise you need to synchronize).

Reply:
also: if you call setStatus(status.FREE); inside the restartOrderProcess() will be called with multple threads ssl.onServerStatusChange(this). This makes it harder to handle all the threads. You export the synchronization part to another class.
Furthermore: if you notify() you know that you will go back to work. But you do not know when. (There could be  50000 years between  serverThread.notify(); and the acually wake up in serverThread.wait(). 
but you change the status immediately. For me it makes sense to say: We stop the state as soon as we start changed the stae.
But you do: We change the state as soon as we know we will change it.

Reply:
I moved it within the synchronized block. I disagree on the 2nd point you raised - when you cann restartOrderProcess on the Server the status should be free as Server is immediately available to accept orders for processing

Reply:
There **is a delay** between notify() and wait().

But i am fine with your solution to, since we know we will work again soon. But keep it consistent. Change the status only from one thread, and make sure everything is synchronized